### ISSUE TITLE
Problematic Handling of Over-Sized Traces (Causing traces to be dropped)

### ISSUE THREAD

**Issue by whited** [DataDog/dd-trace-py#1577]
### TL;DR Problem
I have a DataDog Tracing case which regularly bumps up against the 5MB maximum trace size limit for trace agent payloads.

The tracing library behaviour is technically correct, but I think it should handle very large traces more gracefully.

### More Details
I noticed a long-running gRPC call in a service I am helping to build.  Since we were introducing DataDog tracing, it seemed like a good use case to try out the tracing to determine the problem.

The problem was, the traces for these long-running calls were not appearing.  I eventually tracked down the log entry that explained what was going on:
```
Trace is larger than the max payload size, dropping it
```
(Code reference where the log message is emitted is [here](https://github.com/DataDog/dd-trace-py/blob/c5dceeba31288577b5003d2de06818f1a4378627/ddtrace/api.py#L232))

The dropped trace in question was indeed degenerate.  Based on log timestamps, it took ~10 full seconds to execute.  It had a single service call span, and perhaps thousands of child spans including some absurdly large Django-generated SQL calls.

Tracing is a very useful tool precisely to highlight and detail degenerate cases such as this, but that value is lost when the trace is not present.

### Which version of dd-trace-py are you using?
`0.40.0`

### Which version of the libraries are you using?
n/a

### How can we reproduce your problem?
Build a trace with a **lot** of spans.  The case that I had was a gRPC call that then proceeded to make hundreds (or perhaps thousands) of database calls, all which produced their own child spans.

### What is the result that you get?
A warning log that read `Trace is larger than the max payload size, dropping it`

### What is the result that you expected?

Ideal result would be the trace appearing in the system despite its unreasonably large size.

In the case where the trace is too large to post, perhaps allow for alternative handling:
- eg: Truncating the trace.  Capture the trace id, times, and truncate overly long attributes in the trace.  Favor having the trace show up with some missing details vs not showing up at all.
- eg: Logging the details of the dropped trace.  Some statistics about it, such as byte size, number of attributes, etc. would be helpful.

Configurable max size limits.
- I think I might be able to tweak `Payload.DEFAULT_MAX_PAYLOAD_SIZE` ([ref](https://github.com/DataDog/dd-trace-py/blob/9c4f10061a97564b85e1112454cf095f7ccd13b3/ddtrace/payload.py#L21-L25)) but it doesn't seem like the right way to go about it.  No usages of the `Payload` constructor actually specify `max_payload_size`, so only the default value is ever applied.  A proper end-user configuration that sets this value would be appreciated.

As an aside, the warn log message would be more helpful with additional details.
  - eg: `Trace dropped due to maximum payload size of <max> being exceeded.  Trace size was <size>`

**Comment by denisemauldin** [comment]
I'm also having issues with this. Is there a way I can currently configure our setup so that the trace can be truncated or send in chunks as referenced in the PR?

**Comment by maxmetzger** [comment]
I am having this issue as well. I want to trace very large long-running jobs, so sending a trace in chunks would be ideal.

**Comment by jalaziz** [comment]
This is actually a problem for us too. We have a degenerate request that we're trying to investigate that only affects a subset of users, but Datadog is dropping the traces 😞

**Comment by Kyle-Verhoog** [comment]
Hey everyone, we do have some mitigations in place for this now!

So the tracer writer (as of `v0.50`) can be configured with the environment variables `DD_TRACE_WRITER_BUFFER_SIZE_BYTES`, `DD_TRACE_WRITER_MAX_PAYLOAD_SIZE_BYTES` and `DD_TRACE_WRITER_INTERVAL_SECONDS` as outlined [here](https://ddtrace.readthedocs.io/en/v0.50.3/configuration.html) which can be used to tweak the limits. Please do note that increasing the limits will be a trade-off of consuming more resources for encoding and submitting telemetry.

There is also support coming soon for being able to submit partial traces at a time to the agent which should also help with submitting large numbers of traces. There _is_ a temporary solution with partial flushing (set `DD_TRACER_PARTIAL_FLUSH_ENABLED` and `DD_TRACER_PARTIAL_FLUSH_MIN_SPANS`) **however** this can possibly result in incorrect trace metrics (until the aforementioned support comes) and we only recommend using these if the trace metrics are not valuable for your use-case.

Please let us know if there are any issues with setting these values or if they're insufficient for you use-case!

**Comment by openback** [comment]
Is there any actual documentation on how these variables work? The docs simply list the default values for these, but not what they mean or what implications changing them may have, let alone all of them in tandem
@Kyle-Verhoog

**Comment by Kyle-Verhoog** [comment]
@openback we do have brief descriptions for the variables but it's hard for us to state what the implications might be for an application. eg. setting a much smaller flush interval could result in a big performance regression for applications that handle a lot of requests but might have negligible/no impact on less active apps.

Which variables do you find are not clear or could be expanded on? We're totally open to improving the descriptions in the docs 🙂

**Comment by openback** [comment]
Thanks for the quick reply!

Well looking at `DD_TRACE_WRITER_BUFFER_SIZE_BYTES` alone, I would assume that once a trace reaches that size, that the trace would be flushed and we wouldn't even have an issue with large traces. Or does this value need to be large enough to cover the largest trace?

Likewise with `DD_TRACE_WRITER_MAX_PAYLOAD_SIZE_BYTES`. If we're sending multiple payloads to the agent anyway, why would I need to bump this value?

And I'm not sure what benefit changing `DD_TRACE_WRITER_INTERVAL_SECONDS` would make here either.

**Comment by Kyle-Verhoog** [comment]
> Well looking at DD_TRACE_WRITER_BUFFER_SIZE_BYTES alone, I would assume that once a trace reaches that size, that the trace would be flushed and we wouldn't even have an issue with large traces. Or does this value need to be large enough to cover the largest trace?

It's the max size (because the buffer is infinite) of total encoded traces. So it needs to be large enough to cover the largest trace + whatever other traces occur in the write period (between flushes).

> Likewise with DD_TRACE_WRITER_MAX_PAYLOAD_SIZE_BYTES. If we're sending multiple payloads to the agent anyway, why would I need to bump this value?

This one has a terrible name but hopefully is clarified in the description. It's the max size of a trace that will be added to the buffer.


> And I'm not sure what benefit changing DD_TRACE_WRITER_INTERVAL_SECONDS would make here either.

The flushing strategy is to buffer up to `DD_TRACE_WRITER_BUFFER_SIZE_BYTES` worth of encoded traces every `DD_TRACE_WRITER_INTERVAL_SECONDS` and then send them to the agent. So reducing the interval will mean that more flushes occur to empty the buffer.


We don't really have a section in the docs that explains this mechanism so it might be worth adding a section to our advanced usage docs that explains it.

**Comment by openback** [comment]
So let's say I have the following error.
```
trace (20098348b) larger than payload buffer limit (8388608b), dropping
```
Am I right in thinking that I could just use:
 ```
DD_TRACE_WRITER_MAX_PAYLOAD_SIZE_BYTES=30000000
```
I'm not sure if `DD_TRACE_WRITER_BUFFER_SIZE_BYTES` comes into play here or not

**Comment by Kyle-Verhoog** [comment]
👍 gotcha, just wanted to make sure!
